### ISSUE TITLE
How to use the margin_fully_connected?

### ISSUE THREAD

**Issue by ysc703** [medivhna/TF_Face_Toolbox#1]
Hi, 
I'm really grateful that you share the TF code. 
In the file parallel.py,  the fully_connected layer is used before the loss layer, then the softmax loss and regularization_loss are calculated in the loss_function. I don't find where the margin_fully_connected be used.
But when I use margin_fully_connected instead of fully_connected, it doesn't work.

Thanks in advance!

**Comment by medivhna** [comment]
Thanks for your concern.

The interface of margin_fully_connected layer is slightly different from that of fully_connected. Besides the params where fully_connected needs, you still need to provide **labels** corresponding to input features, **global_step** of your training process, and other hyperparameters for A-softmax loss.

The interface for margin_fully_connected is defined in https://github.com/medivhna/TF_Face_Toolbox/blob/092ec2f243aa1828104190b8ec44b6bf8e621eec/layers/margin_fully_connected.py#L93

I will open a new branch to implement the complete pipeline soon.

**Comment by shangwenxiang** [comment]
@ysc703   Are you successful to use margin_fully_connected.py ?  I do not know what mean of outputs and beta, I just use tensorflow to achieve a-softmax，just like
```
class AngleLinear():
    def __init__(self, in_features, out_features, m=4):
        self.in_features = in_features
        self.out_features = out_features


class AngleLoss():
    def __init__(self, gamma=0):
        self.gamma = gamma
        self.LambdaMin = 4.0
        self.LambdaMax = 1500.0
```

It works well,but just get 98.8 in lfw with casia_webface!

**Comment by ysc703** [comment]
你好
很抱歉最近没注意邮件，issue里也没看到你的问题。
我后来一直没有继续研究了，因为只是抱着学习tf的态度想执行一下，毕竟sphereface采用caffe可以搞定。






At 2017-11-28 18:51:17, "shangwenxiang" <notifications@github.com> wrote:


@ysc703 Are you successful to use margin_fully_connected.py ? I do not know what mean of outputs and beta, I just use tensorflow to achieve a-softmax，just like

class AngleLinear():
    def __init__(self, in_features, out_features, m=4):
        self.in_features = in_features
        self.out_features = out_features


class AngleLoss():
    def __init__(self, gamma=0):
        self.gamma = gamma
        self.LambdaMin = 4.0
        self.LambdaMax = 1500.0


It works well,but just get 98.8 in lfw with casia_webface!
